pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt

```diff
@@ -32,29 +33,34 @@ import org.partiql.pig.domain.model.TypeUniverse
 import org.partiql.pig.util.snakeToCamelCase
 import org.partiql.pig.util.snakeToPascalCase
 
-internal fun TypeUniverse.convertToKTypeUniverse(): KTypeUniverse {
+internal fun TypeUniverse.convertToKTypeUniverse(domains: Set<String>?): KTypeUniverse {
     val allTypeDomains: List<TypeDomain> = this.computeTypeDomains()
 
-    val kotlinTypeDomains: List<KTypeDomain> = allTypeDomains.map { typeDomain ->
-        KTypeDomainConverter(typeDomain).convert()
-    }
+    val kotlinTypeDomains: List<KTypeDomain> = allTypeDomains.filterDomains(domains)
+        .map { typeDomain -> KTypeDomainConverter(typeDomain).convert() }
 
-    val allKTransforms = this.statements.filterIsInstance<Transform>().map { dt ->
-        // TODO: handle properly if one of the calls to .single fails below...
-        val sourceDomain = allTypeDomains.single { it.tag == dt.sourceDomainTag }
-        val destDomain = allTypeDomains.single { it.tag == dt.destinationDomainTag }
-        val difference = sourceDomain.computeTransform(destDomain)
+    val allKTransforms = this.statements.filterIsInstance<Transform>()
 
-        val converter = KTypeDomainConverter(difference)
-        val sourceDomainDifference = converter.convert()
+    val sourceDomains = allKTransforms.map { dt -> allTypeDomains.singleOrNull { it.tag == dt.sourceDomainTag } }
+    val destDomains = allKTransforms.map { dt -> allTypeDomains.singleOrNull { it.tag == dt.destinationDomainTag } }
 
-        KTransform(
-            sourceDomainDifference = sourceDomainDifference,
-            destDomainKotlinName = destDomain.tag.snakeToPascalCase()
-        )
-    }
+    val validKTransforms = sourceDomains.zip(destDomains)
+        .filter { (sourceDomain, destDomain) -> sourceDomain != null && destDomain != null }
+        // sourceDomain and destDomain should both be non-null here, we just checked
+        .filter { (_, destDomain) -> domains?.let { destDomain!!.tag in it } ?: true }
+        .map { (sourceDomain, destDomain) ->
+            val difference = sourceDomain!!.computeTransform(destDomain!!)
+
+            val converter = KTypeDomainConverter(difference)
+            val sourceDomainDifference = converter.convert()
+
+            KTransform(
+                sourceDomainDifference = sourceDomainDifference,
+                destDomainKotlinName = destDomain.tag.snakeToPascalCase()
+            )
+        }
 
-    return KTypeUniverse(kotlinTypeDomains, allKTransforms)
+    return KTypeUniverse(kotlinTypeDomains, validKTransforms)
 }
```

Comment:
Honestly, this function took me too long to figure out how it worked.  Its entirely possible I just haven't had enough coffee this morning.  However, while digging into this in my local IDE I produced this untested version which should be equivalent.  It might be easier to read.  Do with it what you will.

```Kotlin
internal fun TypeUniverse.convertToKTypeUniverse(domains: Set<String>?): KTypeUniverse {
    val allTypeDomains: List<TypeDomain> = this.computeTypeDomains()

    val kotlinTypeDomains: List<KTypeDomain> = allTypeDomains.filterDomains(domains)
        .map { typeDomain -> KTypeDomainConverter(typeDomain).convert() }

    // find all transform statements where the domain is in the `domains` set.
    // or include all transform statements if `domains` is null
    val allTransformStatements = this.statements.filterIsInstance<Transform>()
        .filter { xform -> domains?.let {  xform.destinationDomainTag in domains } ?: true }

    val validKTransforms = allTransformStatements
        .map { xform  ->
            // we should be able to assume these calls to .single wil never fail if this function is called.
            // (missing domains cause a semantic check failure)

            val sourceDomain = allTypeDomains.single { it.tag == xform.sourceDomainTag }
            val destDomain = allTypeDomains.single { it.tag == xform.destinationDomainTag }
            
            val difference = sourceDomain.computeTransform(destDomain)

            val converter = KTypeDomainConverter(difference)
            val sourceDomainDifference = converter.convert()

            KTransform(
                sourceDomainDifference = sourceDomainDifference,
                destDomainKotlinName = destDomain.tag.snakeToPascalCase()
            )
        }

    return KTypeUniverse(kotlinTypeDomains, validKTransforms)
}
```

Reply:
This is less efficient, since it's iterating over all type domains (twice!) inside the `map` statement, when they can be precomputed.  This is what my version does.

Reply:
Otherwise it's functionally the same

Reply:
To be clear I don't really care what you do with that function, but, I just felt it necessary to comment on:

> This is less efficient

Efficiency is a non-goal here while readability is. 

We're talking about a function that executes in <1ms.  This function is also called once during a build that's several minutes long.

Reply:
Readability is subjective - I think my version is more readable since the steps are broken down.  I've incorporated a few changes though.

Reply:
If you have some readability guidelines for this repo, can you point me to them?

Reply:
There are no readability guidelines here.  Of course readability is subjective.  That's why I don't honestly care what you do with the alternative function I posted above.  (Of course, efficiency is a concern for the generated code.)